### PR TITLE
Core: Add a firewall framework + stm32 use case

### PR DESCRIPTION
**Pull request by GseoC**
The need for such framework arises from the fact that there are multiple hardware firewalls implemented across multiple products.  When it comes to firewalls, the purpose mostly stays the same: protect and filter hardware resources. 

Add a firewall framework that offers:
- Firewall device consumers APIs to check, acquire accesses against firewall controllers and firewall reconfiguration as there are cases when a firewall configuration could be dynamic.
- Firewall controllers APIs to register/unregister to the framework and populate their bus.

This firewall framework relies on the _access-controllers_ device tree binding that was acked in the Linux kernel and is present in the Linux-next branch.
How the arguments passed along with this property and how they're used by the controllers to filter the accesses or set configurations is platform-specific AND specific to the firewall controller. Meaning that all the filtering logic should be present in the firewall controller drivers. Firewall consumers do not have the knowledge on the controller and should rely on the device API.

As example, I have reworked and cabled the STM32 ETZPC driver (STM32MP1x series) onto the framework as it implements all of the firewall controllers ops. I also have the RIFSC driver (STM32MP2x series) ready but did not put it in this P-R to avoid pollution. The STM32 ETZPC device is responsible for filtering accesses based on security level, or co-processor isolation for any resource connected to it.

### CODE REVIEW

File: core/drivers/firewall/stm32_etzpc.c [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+			     ETZPC_HWCFGR_NUM_AHB_SEC_SHIFT;
+
+	DMSG("ETZPC revision 0x%02" PRIu8 ", per_sec %u, ahb_sec %u, tzma %u",
+	     io_read8(base + ETZPC_VERR),
+	     ddata->num_per_sec, ddata->num_ahb_sec, ddata->num_tzma);
+}
+
+struct dt_id_attr {
+	/* The effective size of the array is meaningless here */
+	fdt32_t id_attr[1];
```
Comment (GseoC) [inline core/drivers/firewall/stm32_etzpc.c:598]: I'll remove this, it is useless and provokes a warning.

File: core/include/drivers/firewall/firewall.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+#ifndef __DRIVERS_FIREWALL_H
+#define __DRIVERS_FIREWALL_H
+
+#include <compiler.h>
+#include <drivers/firewall/firewall_device.h>
+#include <stddef.h>
+#include <tee_api_defines.h>
+#include <types_ext.h>
```
Comment (etienne-lms) [inline core/include/drivers/firewall/firewall.h:13]: `#include <mm/core_memprot.h>`
`#include <stdbool.h>

File: core/include/drivers/firewall/firewall.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+struct firewall_controller_ops;
```
Comment (etienne-lms) [inline core/include/drivers/firewall/firewall.h:15]: Not used. I think this ops struct could be defined, and referred from `struct firewall_controller`. I would allows driver to use a `static const struct firewall_controller_ops` reference.

File: core/include/drivers/firewall/firewall.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ * @priv:			Private data of the firewall controller
+ * @set_conf:			Set given firewall configuration
+ * @check_access:		Callback used to check access for a consumer
+ *				on a resource against a firewall controller
+ * @acquire_access:		Callback used to acquire access for OP-TEE
+ *				on a resource against a firewall controller
+ * @release_access:		Callback used to release resources taken by a
+ *				consumer when the access was acquired with
+ *				@acquire_access
+ * @check_memory_access:	Callback used to check access for a a consumer
```
Comment (etienne-lms) [inline core/include/drivers/firewall/firewall.h:32]: s/a a/a/

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "core: firewall: add firewall framework".
I wonder if firewall.h and firewall_device.h really need a sub-directory: move them from _core/include/drivers/firewall/_ to _core/include/drivers/_

File: core/drivers/firewall/stm32_etzpc.c [hunk 5]
```diff
@@ -334,6 +329,227 @@ static TEE_Result etzpc_pm(enum pm_op op, unsigned int pm_hint __unused,
...
+	tzma_base = id == ETZPC_TZMA0_ID ? ROM_BASE : SYSRAM_BASE;
+	tzma_size = id == ETZPC_TZMA0_ID ? ROM_SIZE : SYSRAM_SIZE;
+	prot_size = etzpc_do_get_tzma(etzpc_dev, id == ETZPC_TZMA0_ID ? 0 : 1) *
+		    SMALL_PAGE_SIZE;
+
+	DMSG("Acquiring access for TZMA%u, secured from %"PRIxPA" to %"PRIxPA,
+	     id == ETZPC_TZMA0_ID ? 0 : 1, tzma_base, tzma_base + prot_size);
+
+	if (paddr >= tzma_base && paddr < tzma_base + prot_size &&
+	    paddr + size - 1 < tzma_base + tzma_size)
```
Comment (etienne-lms) [inline core/drivers/firewall/stm32_etzpc.c:470]:
```c
	if (core_is_buffer_inside(paddr, size, tzma_base, prot_size))

**Review by etienne-lms — COMMENTED** [review]
Comments for "drivers: stm32_etzpc: new driver to use firewall API".

File: core/arch/arm/plat-stm32mp1/main.c [hunk 6]
```diff
@@ -402,7 +322,7 @@ static TEE_Result init_stm32mp1_drivers(void)
 	return TEE_SUCCESS;
 }
 
-service_init_late(init_stm32mp1_drivers);
+driver_init_late(init_stm32mp1_drivers);
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/arch/arm/plat-stm32mp1/main.c]: Removed that change

**Comment by GseoC** [comment]
Updated with comments addressed.

I added a [modifications] commit that changes a bit the filtering policy of the ETZPC, please have a look at it as well @etienne-lms

File: core/include/drivers/firewall.h [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+#include <types_ext.h>
+
+struct firewall_controller_ops;
+
+/**
+ * struct firewall_controller - Information on firewall controller supplying
+ *				services
+ *
+ * @ops:			Operation handlers
+ * @name:			Name of the firewall controller
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/include/drivers/firewall.h:24]: Ok

File: core/drivers/firewall/stm32_etzpc.c [hunk 8]
```diff
@@ -0,0 +1,707 @@
...
+	assert(valid_tzma_id(etzpc_dev, tzma_id));
+
+	exceptions = etzpc_lock(etzpc_dev);
+
+	io_setbits32(base + ETZPC_TZMA0_SIZE + offset, ETZPC_TZMA0_SIZE_LOCK);
+
+	etzpc_unlock(etzpc_dev, exceptions);
+}
+
+static bool is_tzma_locked(struct etzpc_device *etzpc_dev, uint32_t tzma_id)
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_etzpc.c:281]: Changing decprot as well then.

It really depends on the person IMHO but ok.

File: core/drivers/firewall/stm32_etzpc.c [hunk 9]
```diff
@@ -0,0 +1,707 @@
...
+ * @fconf:	Firewall configurations
+ * @lock:	Access contention
+ */
+struct etzpc_device {
+	struct stm32_etzpc_platdata pdata;
+	struct stm32_etzpc_driver_data *ddata;
+	unsigned int lock;
+};
+
+#if TRACE_LEVEL >= TRACE_DEBUG
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_etzpc.c:121]: mmm ok

File: core/drivers/firewall/stm32_etzpc.c [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+
+/*
+ * STM32 ETPZC acts as a firewall on stm32mp SoC peripheral interfaces and
+ * internal memories. The driver expects a single instance of the controller
+ * in the platform.
+ *
+ * The driver API is defined in header file stm32_etzpc.h.
+ *
+ * Driver registers a PM callback for restoration of the access permissions
+ * when it resumes.
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_etzpc.c:15]: I'm ok to remove it

File: core/drivers/firewall/stm32_etzpc.c [hunk 11]
```diff
@@ -0,0 +1,707 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2015-2021, ARM Limited and Contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/firewall/stm32_etzpc.c:3]: Indeed, thanks

**Comment by GseoC** [comment]
Updated with comments addressed and squashed.

@jenswi-linaro / @jforissier do you want to have a look as well on the generic part?

**Comment by GseoC** [comment]
Fixed a `__maybe_unused` attribute that was misplaced

**Comment by etienne-lms** [comment]
In order to ease reviewers commenting the generic part, maybe it would be worth creating another P-R, maybe [RFC] prefixed, with only the 1 commit of your series, knowing that this current P-R illustrate how an access-controller/firewall device integrates with this new framework.

**Comment by GseoC** [comment]
> In order to ease reviewers commenting the generic part, maybe it would be worth creating another P-R, maybe [RFC] prefixed, with only the 1 commit of your series, knowing that this current P-R illustrate how an access-controller/firewall device integrates with this new framework.

Fine, I'll do that

**Pull request closed without merging**